flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/AvroToBigQuerySchemaTransform.java

```diff
@@ -0,0 +1,456 @@
+            dataType = handleDecimalLogicalType(avroSchema, name);
+            if (dataType == StandardSQLTypeName.NUMERIC
+                    || dataType == StandardSQLTypeName.BIGNUMERIC) {
+                // The precision and scale is also set in the BigQuery Field
+                long precision =
+                        ((LogicalTypes.Decimal) avroSchema.getLogicalType()).getPrecision();
+                long scale = ((LogicalTypes.Decimal) avroSchema.getLogicalType()).getScale();
+                return Field.newBuilder(name, dataType)
+                        .setMode(Field.Mode.REQUIRED)
+                        .setPrecision(precision)
+                        .setScale(scale)
+                        .build();
+            }
+        } else {
+            dataType =
+                    Optional.ofNullable(avroSchema.getProp(LogicalType.LOGICAL_TYPE_PROP))
+                            .map(LOGICAL_AVRO_TYPES_TO_BQ_TYPES::get)
+                            .orElse(AVRO_TYPES_TO_BQ_TYPES.get(avroSchema.getType()));
+        }
+        if (dataType == null) {
+            throw new UnsupportedOperationException(
+                    getErrorMessage(avroSchema.getType().toString(), name));
+        }
+
+        return Field.newBuilder(name, dataType).setMode(Field.Mode.REQUIRED).build();
+    }
+
+    /**
+     * Helper method to handle Avro decimal logical types by determining the appropriate BigQuery
+     * data type based on the precision of the decimal.
+     *
+     * @param avroSchema the Avro schema of the decimal field
+     * @param name the name of the field
+     */
+    private static StandardSQLTypeName handleDecimalLogicalType(
+            org.apache.avro.Schema avroSchema, String name) {
+        long precision = validatePrecisionAndScale(avroSchema, name);
+        long scale = ((LogicalTypes.Decimal) avroSchema.getLogicalType()).getScale();
+
+        if (precision > 0 && precision <= (scale + 29) && scale <= 9) {
+            return StandardSQLTypeName.NUMERIC;
+        } else if (precision > 0 && precision <= (scale + 38) && scale <= 38) {
+            return StandardSQLTypeName.BIGNUMERIC;
+        }
+        LOG.warn(
+                "The precision {} and scale {} of decimal field {} is not supported by BigQuery. Converting the field to type STRING.",
+                precision,
+                scale,
+                name);
+        return StandardSQLTypeName.STRING;
+    }
+
+    /**
+     * Validates the precision and scale of an Avro decimal logical type. Ensures precision and
+     * scale are non-negative and that scale does not exceed precision.
+     *
+     * @param schema the Avro schema of the decimal field
+     * @param name the name of the field
+     * @return the precision of the decimal
+     * @throws IllegalArgumentException if precision or scale are invalid
+     */
+    private static long validatePrecisionAndScale(org.apache.avro.Schema schema, String name) {
```

Comment:
No need to return anything here. Precision can be derived in the caller as well

Reply:
Thanks for pointing this out. Done